gtk4-macros/src/attribute_parser.rs

```diff
@@ -32,14 +32,26 @@ pub fn parse_template_source(input: &DeriveInput) -> Result<TemplateSource> {
         _ => bail!("Invalid meta, specify one of 'file', 'resource', or 'string'"),
     };
 
-    let (ident, v) = parse_attribute(meta)?;
+    let (ident, v) = parse_attribute(source)?;
 
-    match ident.as_ref() {
-        "file" => Ok(TemplateSource::File(v)),
-        "resource" => Ok(TemplateSource::Resource(v)),
-        "string" => Ok(TemplateSource::String(v)),
-        s => bail!("Unknown enum meta {}", s),
-    }
+    let callbacks = meta
```

Comment:
can't we just assume there will always be callbacks? I can't see how useful it's

Reply:
See comment [here](https://github.com/gtk-rs/gtk4-rs/pull/691#issuecomment-964806920)

Reply:
I think it's fine to introduce a breaking change in the next release, we have few others in the pipeline (from gtk-rs-core at least)

Reply:
cc @sdroege , what do you think?

Reply:
I don't understand what this one is about :) But breaking changes are fine if it improves usability, yes.

Reply:
Just for clarity: this would not be breaking for gtk-rs-core, it would only be a breaking change for users of gtk4-rs that use `derive(CompositeTemplate)`.

Reply:
Yeah that's not really a problem. There are other breaking changes for the next release already :)

Reply:
@bilelmoussaoui Explained this to me now. I think not requiring `callbacks` and always going into that code path is the desired behaviour for the next release. If you have callbacks in your UI file and they're not handled then that should complain loudly instead of being silently ignored like in the past.

Reply:
I'm not sure I understand. This won't affect that, what it means is that you would always have to have an empty `impl` with the `template_callbacks` macro on it. We still have no way to complain about missing callbacks unless gtk4-rs does some parsing of the builder XML itself.

Reply:
I guess then we're all talking about different things :)

What @bilelmoussaoui (and I in extension) understood is that if there's a callback inside the UI file that is not mapped to a Rust function then this would cause a panic. That would seem useful. Previously we just ignored callbacks but now we can handle them, so existing code that has callbacks inside the UI file (which were ignored before!) would now panic.

----

But your concern is about requiring `#[gtk::template_callbacks]` on the impl block always now, even if no callbacks are used. That seems suboptimal to me.

Would it be possible to derive the new trait from the existing derive macro, always? And then `#[gtk::template_callbacks]` would only call something on that trait impl to pass the callbacks to it that are then used when creating the scope?

Reply:
>Previously we just ignored callbacks but now we can handle them, so existing code that has callbacks inside the UI file (which were ignored before!) would now panic.

Has that been reported upstream? It seems that the `GtkBuilder` would want to at least throw a warning in that case. I think it would actually be possible to do XML parsing in the proc macro, if we were ok with bringing an extra dependency in there. That way we could check for missing callbacks at compile time, not just panic at run time. That might be neat.

>Would it be possible to derive the new trait from the existing derive macro, always? And then #[gtk::template_callbacks] would only call something on that trait impl to pass the callbacks to it that are then used when creating the scope?

I don't know how this could be done without having `#[gtk::template_callbacks]` generate an additional trait, which is the same problem...

Reply:
> Has that been reported upstream? It seems that the GtkBuilder would want to at least throw a warning in that case. I think it would actually be possible to do XML parsing in the proc macro, if we were ok with bringing an extra dependency in there. That way we could check for missing callbacks at compile time, not just panic at run time. That might be neat. But type checking is another story.

My understanding is that you would run `create_closure` from the builder scope even if the closure doesn't exists which can panic?

Reply:
Hmm, I just checked and I do get a critical error logged from the builder if a handler doesn't exist. `create_closure` isn't supposed to panic on lookup failure, it just returns a `GError` to the builder which is then supposed to handle. I seem to remember Vala having something that would check signals at compile time so maybe that could be copied? (Actually it seems to be [here](https://gitlab.gnome.org/GNOME/vala/-/blob/88801eb4266a8c2d33176739c7f24b6b6cedd774/codegen/valagtkmodule.vala#L376)).

Reply:
> I don't know how this could be done without having #[gtk::template_callbacks] generate an additional trait, which is the same problem...

Me neither. Let's keep it this way for now until someone finds a way :)

Reply:
> Vala having something that would check signals at compile time so maybe that could be copied?

Let's worry about that as a second step IMHO, after this one is merged.

Reply:
One major problem with checking the callbacks during compile time is that the path to the resource file is provided on the `CompositeTemplate` derive macro. It can't be accessed from a separate attribute macro. I don't know a good way to reconcile that without doing something weird like this:

```rust
#[derive(Default, CompositeTemplate)]
#[template(file = "my_widget.ui")]
pub struct MyWidget {
    @callbacks {
        fn my_callback(&self, /* ... */) {
            // ...
        }
    }
}
```

There is also the issue of how to type check. I don't think it can be done effectively without being able to access `GType`s from within the proc macro, which is not going to work. This may be something that is just beyond the ability of Rust macros right now.

Reply:
The callbacks validations at compile time is imho something nice to have in the future, not at all a blocker of this MR and it would integration somehow with the CompositeTemplate to validate that every `#[template_child]` actually exists and is of the correct type. Something for whoever wants to do that work to figure out